File: 0664-strange-printer/0664-strange-printer.cpp
class Solution {
private:
    string s;
    
public:
    int memo[101][101];
    
    int dp(int i, int j) {
        if (j < i) {
            return 0;
        }
        if (i == j) {
            return 1;
        }
        if (memo[i][j] != -1) {
            return memo[i][j];
        }
        int ans = 1 + dp(i + 1, j);
        for (int k = i + 1; k <= j; ++k) {
            if (s[i] == s[k]) {
                ans = min(ans, dp(i + 1, k - 1) + dp(k, j));
            }
        }
        return memo[i][j] = ans;
    }
    
    int strangePrinter(string s) {
        this->s = s;
        memset(memo, -1, sizeof(memo));
        int n = s.size();
        return dp(0, n - 1);
    }
};

File: 0664-strange-printer/0664-strange-printer.py
class Solution:
    def strangePrinter(self, s: str) -> int:
        n = len(s)
        
        @lru_cache(None)
        def dp(l, r):
            if l > r:
                return 0
            
            ans = 1 + dp(l + 1, r)
            
            for k in range(l + 1, r + 1):
                if s[l] == s[k]:
                    ans = min(ans, dp(l + 1, k - 1) + dp(k, r))
                    
            return ans
        
        return dp(0, n - 1)